modem: hl7800: fix IPv6 operation

rerickson1 opened the pull request:
Assign the IPv6 address from the LTE network to the
network iface.
Configure DNS resolver with IPv6 DNS address from the
LTE network.
Fix socket AT commands to account for IPv6 addresses.

fixes #38716

rerickson1 committed:
modem: hl7800: Only reconfig DNS if iface up

In the DNS work callback ensure the iface is up
(on the LTE network) before trying to reconfigure the
DNS resolver with the DNS address.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rerickson1 committed:
modem: hl7800: Add address family config

Add ability to set the IP address family with Kconfig.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rerickson1 committed:
modem: hl7800: ensure IP family set correctly

Ensure that the IP family is synchronized between the PDP
context and the GPRS connection.
There were cases where they could get out of sync.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rerickson1 committed:
modem: hl7800: sync APN for PDN and GPRS connection

Ensure the GPRS connection APN settings match the PDP
context. This it best to ensure proper IP connectivity.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

ycsin commented on drivers/modem/hl7800.c:
```diff
@@ -1707,16 +1726,22 @@ static void dns_work_cb(struct k_work *work)
 	int ret;
 	struct dns_resolve_context *dnsCtx;
-	static const char * const dns_servers_str[] = { ictx.dns_string, NULL };
+	static const char *const dns_servers_str[] = { ictx.dns_v4_string,
```
I remembered that open braces should be the last char in a line?
```c
static const char *const dns_servers_str[] = {
        ictx.dns_v4_string,
        ...
```

rerickson1 replied:
This file was auto-formatted with the .clang-format file using VSCode. So I am going to leave this as is.

ycsin commented on drivers/modem/hl7800.c:
```diff
@@ -311,7 +311,15 @@ static const struct mdm_control_pinconfig pinconfig[] = {
 
-#define SETUP_GPRS_CONNECTION_CMD "AT+KCNXCFG=1,\"GPRS\",\"\",,,\"IPV4V6\""
+#define ADDRESS_FAMILY_IPV4 "IPV4"
+#if defined(CONFIG_MODEM_HL7800_ADDRESS_FAMILY_IPV4V6)
```
Personally from my limited experience I'd prefer
```c
#if IS_ENABLED(CONFIG_*)
```

rather than `#ifdef *` or `#if defined(*)` because `IS_ENABLED` is only true when the CONFIG_* is defined to 1, and the code will compile as intended if the user unintentionally or unknowingly defined `CONFIG_*` = false somewhere and expect the features to be disabled.

Please correct me if I'm wrong

jukkar replied:
Using `IS_ENABLED` is a good habit, but in practice there should be no big difference if one uses `defined`. I recall that there is a check that makes sure user is not able to define `CONFIG_` options in source files. Also one cannot define `CONFIG_*=false` in prj.conf file because Kconfig will make sure that the boolean value needs to be integer.

ycsin replied:
@jukkar Wouldn't `CONFIG_SOMETHING=n` in `prj.conf` or `someboard_defconfig` get compiled into `autoconf.h` as
```c
#define CONFIG_SOMETHING 0
```
, and `#ifdef CONFIG_SOMETHING` will then becomes true?

jukkar replied:
Dunno, but you can certainly try and see what happens. We do not have many invocations of `#if IS_ENABLED()` in the header files atm, so it does not seem be a big concern right now.
`find . -name '*.h' | xargs grep '#if IS_ENABLED'`

ycsin replied:
I checked my app's `autoconf.h` things that are specifically "n" defined in my `prj.conf` aren't being compiled, which is great. I guess I'm just too paranoid.

rerickson1 replied:
Thanks for the discussion. Going to leave this as is.

rlubos reviewed (commented):
Just some nits.

rerickson1 committed:
modem: hl7800: Fix IPv6 operation

Assign the IPv6 address from the LTE network to the
network iface.
Configure DNS resolver with IPv6 DNS address from the
LTE network.
Fix socket AT commands to account for IPv6 addresses.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rerickson1 commented on drivers/modem/hl7800.c:
```diff
@@ -1748,6 +1760,107 @@ char *mdm_hl7800_get_imsi(void)
+ */
+static int hl7800_v6_addr_to_std_v6_addr(const char *src, int src_len, char *dst, int dst_len)
+{
+	int num_sections = 8;
```
agree, this saves some code.

jukkar reviewed (approved):
LGTM
